feat(ui): attestation flow

CodeSandwich opened the pull request:
Part of https://github.com/radicle-dev/radicle-upstream/issues/965. Adds a flow for adding an attestation between a link identity and an Ethereum address.

There's no explicit revocation yet and there are quite a few `TODO`s in the code, but it's enough for an MVP.

CodeSandwich commented on ui/src/attestation/status.ts:
```diff
@@ -0,0 +1,118 @@
+  walletStore: svelteStore.Readable<Wallet>
+): Promise<void> {
+  if (!watching) {
+    watching = true;
```
Fixed

CodeSandwich commented on ui/src/attestation/status.ts:
```diff
@@ -0,0 +1,118 @@
+async function getAttestationStatus(
+  identity: Identity,
+  ethAddress: string,
+  claim_watcher: ClaimWatcher
```
Fixed

CodeSandwich commented on ui/src/attestation/contract.ts:
```diff
@@ -0,0 +1,116 @@
+      .claim(FORMAT_SHA1, payload)
+      .then((ctx: ContractTransaction) =>
+        transaction.add(transaction.claimRadicleIdentity(ctx, urn))
+      );
```
Fixed

CodeSandwich commented on ui/src/identity.ts:
```diff
@@ -16,6 +19,48 @@ export const createIdentity = (
+): Promise<void> => {
+  try {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const { metadata } = session.unsealed()!.identity;
```
Fixed

CodeSandwich commented on ui/src/session.ts:
```diff
@@ -45,6 +45,11 @@ sessionStore.subscribe(data => {
 
+// Return the unseleased session if the session is unsealed, undefined otherwise.
+export const unsealed = (): UnsealedSession | undefined => {
+  return sessionStore.unwrap() as UnsealedSession;
```
Fixed

CodeSandwich commented on ui/src/urn.ts:
```diff
@@ -14,3 +17,34 @@ const urnConstraints = {
   validation.createValidationStore(urnConstraints);
+
+// Parses the identity URN and returns its SHA-1 root hash
+export function parseIdentitySha1(urn: string): Uint8Array {
```
Added

CodeSandwich commented on ui/src/attestation/contract.ts:
```diff
@@ -0,0 +1,116 @@
+    onClaimed: (claimed?: Uint8Array) => void
+  ): Promise<void> {
+    const removedFilter = this.contract.filters.Claimed(null);
+    await this.contract.removeAllListeners(removedFilter);
```
Fixed

CodeSandwich commented on ui/src/identity.ts:
```diff
@@ -16,6 +19,48 @@ export const createIdentity = (
+  try {
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const { metadata } = session.unsealed()!.identity;
+    await api.put<Metadata, void>("identities", {
```
Fixed

geigerzaehler commented on ui/src/urn.test.ts:
```diff
@@ -0,0 +1,78 @@
+  }
+
+  it("parses a valid identity URN", () => {
+    expectSuccess(
+      "rad:git:hnrkr5f1pdx7jysj5wkep9ci3pfaztfrgt7ay",
+      "0x4d964d1bfa90593ba290dfb2b969717894868f70"
+    );
```
I’d prefer to inline the `expectSuccess` function since it is only used one but I’m also ok if you want to leave it like this. In general I really started to like this pattern of extracting the test body :sparkles:.

CodeSandwich replied:
I'd prefer keeping it separate. We're wasting just a few lines of code, but we're keeping the whole series of tests clear and consistent, we're also encouraging writing more success cases if they are needed.

geigerzaehler commented on ui/src/attestation/contract.ts:
```diff
@@ -0,0 +1,114 @@
+    txHash: string,
+    address: string
+  ): Promise<Uint8Array> {
+    const SKIP = false;
+    if (SKIP) {
+      return new Uint8Array();
+    }
```
This shouldn’t be here, right?

CodeSandwich replied:
Removed

CodeSandwich commented on ui/src/attestation/contract.ts:
```diff
@@ -0,0 +1,114 @@
+    if (tx.from != address) {
+      throw new Error("Claim transaction sent by an invalid address");
+    }
+    if (tx.to != this.contract.address) {
```
Fixed in all places

CodeSandwich commented:
Fixed the issue with the app not noticing a complete attestation after reconnecting to a different wallet with a different address (2 little missing `await`s). Fixed the ugly transaction details card (thank you Rudolfs!).

NunoAlexandre reviewed (dismissed):
nice :surfing_man:

NunoAlexandre commented on ui/src/attestation/status.ts:
```diff
@@ -0,0 +1,124 @@
+  watching = true;
+
+  const wallet = svelteStore.get(walletStore);
+  const claimWatcher = new ClaimWatcher(
```
we should unwatch this :thinking:

CodeSandwich replied:
The `ClaimWatcher` unwatches and rewatches automatically based on the wallet address changes. We would need to explicitly unwatch only if we wanted to stop watching entirely and stop the polling loop below. It may make sense in the future, but now we have an infinite loop singleton always running in the back.

rudolfs commented on ui/src/attestation/contract.ts:
```diff
@@ -0,0 +1,110 @@
+      ethers.BigNumberish,
+      ethers.BytesLike
+    ] = this.contract.interface.decodeFunctionData("claim", tx.data);
+    if (FORMAT_SHA1.eq(format) == false) {
```
Unless there's a good reason not to, we should stick to the convention of only using strict equality operators: `!==` and `===`. Noticed also in a couple of other places in this changeset.

From [JS the good parts]( https://www.goodreads.com/book/show/2998152-javascript):
> JavaScript has two sets of equality operators: === and !==, and their evil twins == and !=. The good ones work the way you would expect. If the two operands are of the same type and have the same value, then === produces true and !== produces false. The evil twins do the right thing when the operands are of the same type, but if they are of different types, they attempt to coerce the values. the rules by which they do that are complicated and unmemorable.

CodeSandwich replied:
Fixed

NunoAlexandre reviewed (dismissed):
Great work, Igor :star2:

rudolfs reviewed (approved):
Tested ACK
